### PR TITLE
Implement `ShadowRoot.innerHtml` attribute

### PR DESCRIPTION
**Pull request by simonwuelker**
https://html.spec.whatwg.org/multipage/#dom-shadowroot-innerhtml

Unfortunately this exposes two new crashes and a failure, all of them because they use `root.innerHtml = ...` to set up the test, which previously did not work correctly.

[try run](https://github.com/simonwuelker/servo/actions/runs/11962782911)

Part of #34318

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] There are tests for these changes

### CODE REVIEW

File: components/script/dom/documentfragment.rs [hunk 1]
```diff
@@ -60,6 +60,38 @@ impl DocumentFragment {
...
+    /// <https://html.spec.whatwg.org/multipage/#fragment-serializing-algorithm-steps>
+    pub fn fragment_serialization_algorithm(&self, require_well_formed: bool) -> DOMString {
+        // Step 1. Let context document be node's node document.
+        let context_document = document_from_node(self);
+
+        // Step 2. If context document is an HTML document, return the result of HTML fragment serialization algorithm
+        // with node, false, and « ».
+        let mut writer = vec![];
+        if context_document.is_html_document() {
+            html5ever::serialize(
```
Comment (jdm) [inline components/script/dom/documentfragment.rs:73]: Can we make https://github.com/servo/servo/blob/main/components/script/dom/element.rs#L1330 a helper we can reuse here instead?

File: components/script/dom/documentfragment.rs [hunk 2]
```diff
@@ -60,6 +60,38 @@ impl DocumentFragment {
...
+            )
+            .expect("Cannot serialize element");
+
+            return DOMString::from(String::from_utf8(writer).unwrap());
+        }
+
+        // Step 3. Return the XML serialization of node given require well-formed.
+        // TODO: xml5ever doesn't seem to want require_well_formed
+        let _ = require_well_formed;
+        xml5ever::serialize::serialize(
```
Comment (jdm) [inline components/script/dom/documentfragment.rs:86]: Similarly with https://github.com/servo/servo/blob/main/components/script/dom/element.rs#L1347 ?

**Review by jdm — APPROVED** [review]

**Comment by simonwuelker** [comment]
I used the helper methods in `element.rs` as suggested, but moved them (and the fragment serialization algorithm itself) into member functions of `Node`. While a Node is not logically HTML or XML, these functions really only want node functionality and the fragment serialization algorithm defined for every child-interface of Node.

File: components/script/dom/node.rs [hunk 3]
```diff
@@ -2448,6 +2449,52 @@ impl Node {
...
+
+    /// <https://html.spec.whatwg.org/multipage/#fragment-serializing-algorithm-steps>
+    pub fn fragment_serialization_algorithm(&self, require_well_formed: bool) -> DOMString {
+        // Step 1. Let context document be node's node document.
+        let context_document = document_from_node(self);
+
+        // Step 2. If context document is an HTML document, return the result of HTML fragment serialization algorithm
+        // with node, false, and « ».
+        if context_document.is_html_document() {
+            return self.xml_serialize(html_serialize::TraversalScope::ChildrenOnly(None));
```
Comment (jdm) [inline components/script/dom/node.rs:2490]:
```suggestion
            return self.hml_serialize(html_serialize::TraversalScope::ChildrenOnly(None));
```

Comment (simonwuelker) [reply components/script/dom/node.rs]: Oops

**Pull request merged**